distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering/Natural_Order.enso

```diff
@@ -17,62 +20,66 @@ import Standard.Base.Data.Ordering.Vector_Lexicographic_Order
+                character = Text_Utils.get_chars substring . at 0
+                if (is_digit character).not then next else
+                    @Tail_Call find_number text new_next iter
+
+        end_index = find_number text next iter
+        substring = Text_Utils.substring text prev n.abs
+        
+        ## TODO [RW] Currently there is no `Integer.parse` method, so we
                parse a decimal and convert it to an integer. Once
                https://www.pivotaltracker.com/story/show/181176522 is
-               implemented, this should be changed to use `Integer.parse`.
-            value = Decimal.parse part . floor
-            Natural_Ordering_Numeric_Part value part
-        False ->
-            Natural_Ordering_Alphabetical_Part part
-
-## PRIVATE
-type Natural_Ordering_Part
-    ## PRIVATE
-    type Natural_Ordering_Alphabetical_Part (text : Text)
-
-    ## PRIVATE
-    type Natural_Ordering_Numeric_Part (value : Integer) (original_text : Text)
-
-    ## PRIVATE
-    compare_to : Natural_Ordering_Part -> Ordering
-    compare_to that = case Pair this that of
-        Pair (Natural_Ordering_Alphabetical_Part text1) (Natural_Ordering_Alphabetical_Part text2) ->
-            text1 . compare_to text2
-        Pair (Natural_Ordering_Numeric_Part value1 original1) (Natural_Ordering_Numeric_Part value2 original2) ->
-            value_ordering = value1.compare_to value2
-            case value_ordering of
-                Ordering.Equal -> original1 . compare_to original2
-                _ -> value_ordering
-        Pair (Natural_Ordering_Numeric_Part _ _) (Natural_Ordering_Alphabetical_Part _) ->
-            Ordering.Less
-        Pair (Natural_Ordering_Alphabetical_Part _) (Natural_Ordering_Numeric_Part _ _) ->
-            Ordering.Greater
-
-    to_text : Text
-    to_text = case this of
-        Natural_Ordering_Alphabetical_Part text -> text.to_text
-        Natural_Ordering_Numeric_Part v o -> [v, o].to_text
+               implemented, this should be changed to use `Integer.parse`.        
+        decimal = Decimal.parse substring . floor
+
+        next_index = if end_index < 0 then -1 else iter.current
+        [substring, decimal, end_index, next_index]
+
+
+    ## Ordering: Nothing < Number < Text
+    order prev1 next1 prev2 next2 =
+        case (next1 == -1) of
+            True ->
+                if (next2 == -1) then Ordering.Equal else Ordering.Less
+            False ->
+                if (next2 == -1) then Ordering.Greater else
+                    substring1 = Text_Utils.substring text1 prev1 next1
+                    character1 = Text_Utils.get_chars substring1 . at 0
+
+                    substring2 = Text_Utils.substring text2 prev2 next2
+                    character2 = Text_Utils.get_chars substring2 . at 0
+
+                    case (is_digit character1) of
+                        True ->
+                            if (is_digit character2).not then Ordering.Less else
```

Comment:
I would also suggest using the `case Pair ... ... of` construct here. This may be subjective, but IMO it is more readable.